File: samples/python/authenticate.py
"""
License: Apache 2.0. See LICENSE file in root directory.
Copyright(c) 2020-2021 RealSense, Inc. All Rights Reserved.

Sample demonstrating face authentication.
Auto-discovers the device and runs a single authentication.
"""

import rsid_py

FACE_LANDMARK_NAMES = ["left_eye", "right_eye",
                       "nose", "left_mouth", "right_mouth"]


def on_result(result, user_id):
    if result == rsid_py.AuthenticateStatus.Success:
        print(f'\tAuthenticated user: {user_id}')
    else:
        print(f'\tAuthentication failed with status: {result}')


def on_hint(hint, score):
    print(f'\tReceived hint {hint} with score {score:.2f}')


def on_landmarks(landmarks, _timestamp):
    for i, lm in enumerate(landmarks):
        parts = [f'{FACE_LANDMARK_NAMES[j]}=({lm.lm_x[j]}, {lm.lm_y[j]})' for j in range(
            len(FACE_LANDMARK_NAMES))]
        print(f'\tFace {i}: {" ".join(parts)}')


def on_face_distances(distances, _timestamp):
    for i, d in enumerate(distances):
        print(f'\tFace {i} distance: {d:.2f} cm')


def on_faces(faces, _timestamp):
    for i, f in enumerate(faces):
        print(f'\tFace {i} detected at ({f.x}, {f.y}) with size {f.w}x{f.h}')


if __name__ == '__main__':
    devices = rsid_py.discover_devices()
    if not devices:
        print("Error: No RealSenseID device detected.")
        exit(1)
    if len(devices) > 1:
        print("Error: Multiple devices detected. Please connect only one.")
        exit(1)
    device = devices[0]
    print(f"\tUsing device on port {device.serial_port}")

    with rsid_py.FaceAuthenticator(device.device_type, device.serial_port) as f:
        f.authenticate(
            on_faces=on_faces,
            on_hint=on_hint,
            on_landmarks=on_landmarks,
            on_face_distances=on_face_distances,
            on_result=on_result,
        )

File: samples/cpp/authenticate.cc
// License: Apache 2.0. See LICENSE file in root directory.
// Copyright(c) 2020-2021 RealSense, Inc. All Rights Reserved.

#include "RealSenseID/FaceAuthenticator.h"
#include "RealSenseID/DiscoverDevices.h"
#include <iostream>
#include <stdio.h>

class MyAuthClbk : public RealSenseID::AuthenticationCallback
{
public:
    void OnResult(const RealSenseID::AuthenticateStatus status, const char* user_id, short score) override
    {
        if (status == RealSenseID::AuthenticateStatus::Success)
            std::cout << "Authenticated " << user_id << std::endl;
    }

    void OnHint(const RealSenseID::AuthenticateStatus hint, float frameScore) override
    {
        std::cout << "OnHint " << hint << std::endl;
    }

    void OnFaceDetected(const std::vector<RealSenseID::FaceRect>& faces, const unsigned int ts) override
    {
        for (auto& face : faces)
        {
            printf("** Detected face %u,%u %ux%u (timestamp %u)\n", face.x, face.y, face.w, face.h, ts);
        }
    }

    void OnLandmarksDetected(const std::vector<RealSenseID::FaceLandmarks>& landmarks, const unsigned int ts) override
    {
        for (auto& lms : landmarks)
        {
            printf("** Detected landmarks (timestamp %u)\n", ts);
            for (size_t l = 0; l < NUM_FACE_LANDMARKS; l++)
            {
                printf("    x[%zu]=%d, y[%zu]=%d\n", l, lms.lm_x[l], l, lms.lm_y[l]);
            }
        }
    }

    void OnPersonDetected(const std::vector<RealSenseID::PersonRect>& persons, unsigned int ts) override
    {
        for (const auto& person : persons)
        {
            printf("** Detected person (id %u) %u,%u %ux%u (timestamp %u) score=%.3f\n", person.id, person.x, person.y, person.w, person.h,
                   ts, person.score);
        }
    }

    void OnBarcodeDecoded(const std::vector<std::string>& barcodes, unsigned int ts) override
    {
        for (const auto& barcode : barcodes)
        {
            printf("** Detected barcode %s (timestamp %u)\n", barcode.c_str(), ts);
        }
    }
};

int main()
{
    auto devices = RealSenseID::DiscoverDevices();
    if (devices.empty())
    {
        std::cout << "No device detected" << std::endl;
        return 1;
    }
    auto& device = devices.front();
    std::cout << "Using device on port " << device.serialPort << std::endl;

    RealSenseID::FaceAuthenticator authenticator(device.deviceType);
    auto status = authenticator.Connect({device.serialPort});
    if (status != RealSenseID::Status::Ok)
    {
        std::cout << "Failed connecting with status " << status << std::endl;
        return 1;
    }
    MyAuthClbk auth_clbk;
    authenticator.Authenticate(auth_clbk);
}
